src/session.rs

```diff
@@ -0,0 +1,73 @@
+use crate::{
+    constants::{algorithm::HashingAlgorithm, types::session::SessionType},
+    handles::SessionHandle,
+    Error, Result, WrapperErrorKind,
+};
+use log::error;
+use std::convert::{From, TryFrom};
+
+/// Enum representing the different types of sessions.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum Session {
+    Hmac(HmacSession),
+    Policy(PolicySession),
+    Trial(TrialSession),
+}
+
+/// Macro for implementing specific session types
+macro_rules! impl_session_data_type {
+    ($specific_session_type:ident, Session::$session_enum_value:ident, SessionType::$session_type_value:ident) => {
+        #[derive(Debug, Copy, Clone, Eq, PartialEq)]
+        pub struct $specific_session_type {
+            handle: SessionHandle,
+            auth_hash: HashingAlgorithm,
+        }
+
+        impl $specific_session_type {
+            pub fn new(
+                handle: SessionHandle,
+                auth_hash: HashingAlgorithm,
+            ) -> $specific_session_type {
+                $specific_session_type { handle, auth_hash }
+            }
+
+            pub fn handle(&self) -> SessionHandle {
+                self.handle
+            }
+
+            pub fn auth_hash(&self) -> HashingAlgorithm {
+                self.auth_hash
+            }
+
+            pub const fn session_type() -> SessionType {
+                SessionType::$session_type_value
+            }
+        }
+
+        impl From<$specific_session_type> for Session {
+            fn from(specific_session: $specific_session_type) -> Session {
+                Session::$session_enum_value(specific_session)
+            }
+        }
+
+        impl TryFrom<Session> for $specific_session_type {
+            type Error = Error;
+            fn try_from(session: Session) -> Result<$specific_session_type> {
+                match session {
+                    Session::$session_enum_value(val) => Ok(val),
+                    _ => {
+                        error!(
+                            "Error to convert session into {}",
+                            std::stringify!($specific_session_type)
+                        );
+                        Err(Error::local_error(WrapperErrorKind::InvalidParam))
+                    }
+                }
+            }
+        }
```

Comment:
Well it when I think about, no it is not needed. The only way to create a session is by calling StartAuthSession. I think I was taking to much inspiration from the TpmHandles but they are completely different. This should be removed.

Reply:
Hrrm, on second thought.

I was thinking that if a context method would require a special kind of session then there have to be a way to convert the result from StartAuthSession into that special kind and providing this conversion the way of doing it will be pretty nice:
```rust
        let trial_session = context
            .start_auth_session(
                ESYS_TR_NONE,
                ESYS_TR_NONE,
                None,
                SessionType::Trial,
                Cipher::aes_256_cfb(),
                HashingAlgorithm::Sha256,
            )
            .and_then(TrialSession::try_from)
            .unwrap(); // cannot fail
```

Reply:
I think I will keep this.